### ISSUE TITLE
currency_m60.py

### ISSUE THREAD

**Issue by akutsuwakutsu** [akutsuwakutsu/.github.io#20]
import pandas as pd
import numpy as np
import lightgbm as lgb
import os

# ========================================
# 設定
# ========================================

TP = 0.50   # 50 pips
SL = 0.50   # 50 pips
ENTRY_THRESHOLD = 0.60

TRAIN_YEARS_BACK = 2
TEST_YEARS = list(range(2017, 2026))

# 対象通貨ペア
PAIRS = [
    "USDJPY", "EURUSD", "GBPUSD", "USDCAD",
    "USDSEK", "USDCHF", "EURJPY", "AUDJPY", "NZDJPY"
]

# 時間枠（分）
TIMEFRAMES = [5, 15, 30, 60]

# ========================================
# データ読み込み・リサンプリング
# ========================================

def load_1m_csv(filename):
    """1分足CSVを読み込み、重複行を削除"""
    df = pd.read_csv(
        filename,
        header=None,
        names=["Date", "Time", "Open", "High", "Low", "Close", "Volume"]
    )
    df["Datetime"] = pd.to_datetime(df["Date"] + " " + df["Time"])
    df = df.set_index("Datetime")
    df = df[~df.index.duplicated(keep='first')]
    df = df[["Open", "High", "Low", "Close", "Volume"]]
    return df

def resample_ohlc(df_1m, timeframe_minutes):
    """1分足を指定時間足にリサンプリング"""
    freq = f"{timeframe_minutes}min"
    df_res = df_1m.resample(freq).agg({
        "Open": "first",
        "High": "max",
        "Low": "min",
        "Close": "last",
        "Volume": "sum"
    })
    return df_res.dropna()

# ========================================
# 特徴量作成（60分足基準）
# ========================================

def make_features(pair_data_dict):
    """
    pair_data_dict: {通貨ペア名: 1分足DataFrame}
    各通貨ペアについて、マルチタイムフレームMA20特徴量を作成し、
    60分足インデックスに揃えたDataFrameを返す。
    """
    # USDJPYの60分足を基準インデックスにする
    usdjpy_1m = pair_data_dict["USDJPY"]
    usdjpy_60m = resample_ohlc(usdjpy_1m, 60)
    base_index = usdjpy_60m.index
    base_close = usdjpy_60m["Close"]

    all_features = pd.DataFrame(index=base_index)

    for pair in PAIRS:
        df_1m = pair_data_dict[pair]

        for tf in TIMEFRAMES:
            df_tf = resample_ohlc(df_1m, tf)
            if df_tf.empty:
                continue

            ma20 = df_tf["Close"].rolling(20).mean()
            slope = ma20.diff()

            # 各時間足バーの終了時刻
            tf_end_time = df_tf.index + pd.Timedelta(minutes=tf)
            ma20_by_end = pd.Series(ma20.values, index=tf_end_time)
            slope_by_end = pd.Series(slope.values, index=tf_end_time)

            # 60分足の各バーの終了時刻
            sixty_min_end = base_index + pd.Timedelta(minutes=60)

            # 60分足終了時点で利用可能な最新のMA20・傾きを取得
            ma20_aligned = pd.merge_asof(
                pd.DataFrame({"end": sixty_min_end}),
                ma20_by_end.rename("ma20"),
                left_on="end",
                right_index=True,
                direction="backward"
            )["ma20"]
            ma20_aligned.index = base_index

            slope_aligned = pd.merge_asof(
                pd.DataFrame({"end": sixty_min_end}),
                slope_by_end.rename("slope"),
                left_on="end",
                right_index=True,
                direction="backward"
            )["slope"]
            slope_aligned.index = base_index

            dist_name = f"{pair}_ma20_{tf}m_distance"
            slope_name = f"{pair}_ma20_{tf}m_slope"

            # 距離 = 現在の60分足終値 - その時間足のMA20
            all_features[dist_name] = base_close - ma20_aligned
            all_features[slope_name] = slope_aligned

    all_features = all_features.replace([np.inf, -np.inf], np.nan)
    return all_features

# ========================================
# ラベル作成（60分足エントリー・TP/SL50）
# ========================================

def make_labels(df_60m, df_1m, tp, sl):
    """
    各60分足に対し、次の60分足始値でエントリーした結果のラベルを作成。
    同時到達・No-ResultはNaN。
    """
    labels = []
    counts = {"win": 0, "loss": 0, "simultaneous": 0, "no_result": 0, "skip": 0}

    one_min_times = df_1m.index.values
    one_min_high = df_1m["High"].values
    one_min_low = df_1m["Low"].values

    sixty_min_times = df_60m.index.values
    sixty_min_open = df_60m["Open"].values

    n = len(df_60m)

    for i in range(n):
        if i == n - 1:
            labels.append(np.nan)
            counts["skip"] += 1
            continue

        entry_time = sixty_min_times[i + 1]
        entry_price = sixty_min_open[i + 1]

        entry_idx = np.searchsorted(one_min_times, entry_time)
        if entry_idx >= len(one_min_times):
            labels.append(np.nan)
            counts["skip"] += 1
            continue

        tp_price = entry_price + tp
        sl_price = entry_price - sl

        outcome = None
        for j in range(entry_idx, len(one_min_times)):
            h = one_min_high[j]
            l = one_min_low[j]
            hit_tp = h >= tp_price
            hit_sl = l <= sl_price

            if hit_tp and hit_sl:
                outcome = "simultaneous"
                break
            elif hit_tp:
                outcome = "win"
                break
            elif hit_sl:
                outcome = "loss"
                break

        if outcome == "win":
            labels.append(1)
            counts["win"] += 1
        elif outcome == "loss":
            labels.append(0)
            counts["loss"] += 1
        elif outcome == "simultaneous":
            labels.append(np.nan)
            counts["simultaneous"] += 1
        else:
            labels.append(np.nan)
            counts["no_result"] += 1

    return pd.Series(labels, index=df_60m.index), counts

# ========================================
# バックテスト（1ポジション管理・60分足）
# ========================================

def backtest_year(predictions, df_60m, df_1m, threshold, tp, sl):
    counts = {"win": 0, "loss": 0, "simultaneous": 0, "no_result": 0}
    trades = []

    pred_sorted = predictions.sort_index()
    pred_times = pred_sorted.index.values
    pred_values = pred_sorted.values

    sixty_min_times = df_60m.index.values
    sixty_min_open = df_60m["Open"].values

    i = 0
    in_position = False
    resolution_time = None

    while i < len(pred_sorted):
        current_time = pred_times[i]

        if in_position and current_time <= resolution_time:
            i += 1
            continue
        if in_position and current_time > resolution_time:
            in_position = False

        prob = pred_values[i]
        if prob < threshold:
            i += 1
            continue

        try:
            pos = np.where(sixty_min_times == current_time)[0][0]
        except IndexError:
            i += 1
            continue

        if pos + 1 >= len(sixty_min_times):
            i += 1
            continue

        entry_time = sixty_min_times[pos + 1]
        entry_price = sixty_min_open[pos + 1]

        # 1分足でTP/SL判定
        one_min_times = df_1m.index.values
        one_min_high = df_1m["High"].values
        one_min_low = df_1m["Low"].values

        entry_idx = np.searchsorted(one_min_times, entry_time)
        if entry_idx >= len(one_min_times):
            i += 1
            continue

        tp_price = entry_price + tp
        sl_price = entry_price - sl

        outcome = None
        for j in range(entry_idx, len(one_min_times)):
            h = one_min_high[j]
            l = one_min_low[j]
            hit_tp = h >= tp_price
            hit_sl = l <= sl_price

            if hit_tp and hit_sl:
                outcome = "simultaneous"
                resolution_time = one_min_times[j]
                break
            elif hit_tp:
                outcome = "win"
                resolution_time = one_min_times[j]
                break
            elif hit_sl:
                outcome = "loss"
                resolution_time = one_min_times[j]
                break

        if outcome == "win":
            counts["win"] += 1
            trades.append({"entry_time": entry_time, "outcome": 1, "probability": prob})
        elif outcome == "loss":
            counts["loss"] += 1
            trades.append({"entry_time": entry_time, "outcome": 0, "probability": prob})
        elif outcome == "simultaneous":
            counts["simultaneous"] += 1
        else:
            counts["no_result"] += 1
            resolution_time = one_min_times[-1]

        in_position = True
        i += 1

    total_trades = counts["win"] + counts["loss"]
    win_rate = (counts["win"] / total_trades * 100) if total_trades > 0 else 0
    profit = counts["win"] * tp - counts["loss"] * sl

    return counts, win_rate, profit, pd.DataFrame(trades)

# ========================================
# 年次データ処理（キャッシュ付き）
# ========================================

data_cache = {}

def get_year_data(year):
    """9通貨ペアの1分足を読み込み、60分足特徴量を計算して返す"""
    if year in data_cache:
        return data_cache[year]

    print(f"\n===== {year}年 データ読み込み・特徴量作成 =====")

    pair_data = {}
    for pair in PAIRS:
        filename = f"{pair}_{year}_all.csv"
        if not os.path.exists(filename):
            print(f"ファイルが見つかりません: {filename}")
            data_cache[year] = None
            return None
        pair_data[pair] = load_1m_csv(filename)

    # USDJPYの60分足・1分足
    usdjpy_1m = pair_data["USDJPY"]
    usdjpy_60m = resample_ohlc(usdjpy_1m, 60)

    # 特徴量
    features = make_features(pair_data)

    # ラベル
    labels, label_counts = make_labels(usdjpy_60m, usdjpy_1m, TP, SL)

    data = {
        "df_1m": usdjpy_1m,
        "df_60m": usdjpy_60m,
        "features": features,
        "labels": labels,
        "label_counts": label_counts
    }
    data_cache[year] = data
    return data

# ========================================
# メイン処理
# ========================================

print("========================================")
print("60分足トレード・マルチタイムフレームMA20特徴量モデル")
print(f"通貨ペア数: {len(PAIRS)}")
print(f"特徴量数: 9 × 2 × 4 = 72")
print(f"TP: {TP} pips / SL: {SL} pips")
print(f"エントリー閾値: {ENTRY_THRESHOLD}")
print(f"学習: 直近{TRAIN_YEARS_BACK}年 → 翌年テスト")
print("========================================")

yearly_results = {}   # test_year -> {counts, win_rate, profit, trades_df}
all_importances = []
all_trades = []       # 全テスト年の取引結果（確率帯分析用）

for test_year in TEST_YEARS:
    train_years = list(range(test_year - TRAIN_YEARS_BACK, test_year))
    print(f"\n########## {train_years}年学習 → {test_year}年テスト ##########")

    # 学習データ準備
    train_features_list = []
    train_labels_list = []
    for year in train_years:
        data = get_year_data(year)
        if data is None:
            continue
        train_features_list.append(data["features"])
        train_labels_list.append(data["labels"])

    if len(train_features_list) == 0:
        print("学習データなし")
        continue

    X_train = pd.concat(train_features_list)
    y_train = pd.concat(train_labels_list)

    train = X_train.copy()
    train["target"] = y_train
    train = train.dropna()

    FEATURES = list(X_train.columns)
    X_train_clean = train[FEATURES]
    y_train_clean = train["target"].astype(int)

    # テストデータ
    test_data = get_year_data(test_year)
    if test_data is None:
        continue

    X_test = test_data["features"].dropna()
    test_index = X_test.index

    # LightGBM デフォルトモデル
    model = lgb.LGBMClassifier(
        objective="binary",
        random_state=42,
        verbosity=-1
    )
    model.fit(X_train_clean, y_train_clean)

    all_importances.append(model.feature_importances_)

    # 予測
    probabilities = model.predict_proba(X_test)[:, 1]
    predictions = pd.Series(probabilities, index=test_index)

    # バックテスト
    counts, win_rate, profit, trades_df = backtest_year(
        predictions,
        test_data["df_60m"],
        test_data["df_1m"],
        ENTRY_THRESHOLD,
        TP,
        SL
    )

    yearly_results[test_year] = {
        "counts": counts,
        "win_rate": win_rate,
        "profit": profit,
        "trades_df": trades_df
    }

    # 確率帯分析用に取引結果を保存
    if not trades_df.empty:
        trades_df["year"] = test_year
        all_trades.append(trades_df)

    print(f"  {test_year}年: 勝ち={counts['win']}, 負け={counts['loss']}, "
          f"勝率={win_rate:.2f}%, 損益={profit:+.3f} ({profit*100:+.1f} pips)")

# ========================================
# ① 総合結果
# ========================================

print("\n\n========================================")
print("① 総合結果（全テスト年合計）")
print("========================================")

total_win = sum(yearly_results[y]["counts"]["win"] for y in yearly_results)
total_loss = sum(yearly_results[y]["counts"]["loss"] for y in yearly_results)
total_profit = sum(yearly_results[y]["profit"] for y in yearly_results)
total_trades = total_win + total_loss
overall_win_rate = (total_win / total_trades * 100) if total_trades > 0 else 0

print(f"総勝ち: {total_win}")
print(f"総負け: {total_loss}")
print(f"総損益: {total_profit:+.3f} ({total_profit*100:+.1f} pips)")
print(f"勝率: {overall_win_rate:.2f}%")

# ========================================
# ② 年別成績
# ========================================

print("\n\n========================================")
print("② 年別成績")
print("========================================")
print(f"{'年':<6} {'勝ち':>6} {'負け':>6} {'勝率':>8} {'損益(pips)':>12}")
print("-" * 45)

for test_year in TEST_YEARS:
    if test_year in yearly_results:
        c = yearly_results[test_year]["counts"]
        win = c["win"]
        loss = c["loss"]
        wr = yearly_results[test_year]["win_rate"]
        profit_pips = yearly_results[test_year]["profit"] * 100
        print(f"{test_year:<6} {win:>6} {loss:>6} {wr:>7.2f}% {profit_pips:>+11.1f}")
    else:
        print(f"{test_year:<6}  データなし")

print("-" * 45)

# ========================================
# ③ 特徴量重要度ランキング
# ========================================

print("\n\n========================================")
print("③ 特徴量重要度ランキング（全テスト年平均）")
print("========================================")

avg_importance = np.mean(all_importances, axis=0)

importance_df = pd.DataFrame({
    "feature": FEATURES,
    "importance": avg_importance
}).sort_values("importance", ascending=False)

print(f"{'順位':<4} {'特徴量':<35} {'重要度':>10}")
print("-" * 55)

for rank, (_, row) in enumerate(importance_df.iterrows(), start=1):
    print(f"{rank:<4} {row['feature']:<35} {row['importance']:>10.0f}")

# ========================================
# ④ 予測確率帯別 実績勝率
# ========================================

print("\n\n========================================")
print("④ 予測確率帯別 実績勝率（全テスト年合算）")
print("========================================")

if all_trades:
    combined_trades = pd.concat(all_trades, ignore_index=True)

    # 確率帯を0.05刻みで作成（0.50から0.70以上）
    bins = [0.50, 0.55, 0.60, 0.65, 0.70, np.inf]
    labels = ["0.50～0.55", "0.55～0.60", "0.60～0.65", "0.65～0.70", "0.70～"]

    combined_trades["prob_bin"] = pd.cut(
        combined_trades["probability"],
        bins=bins,
        labels=labels,
        right=False
    )

    bin_summary = combined_trades.groupby("prob_bin", observed=False).agg(
        count=("outcome", "size"),
        wins=("outcome", "sum")
    )
    bin_summary["win_rate"] = bin_summary["wins"] / bin_summary["count"] * 100

    print(f"{'予測確率':<12} {'件数':>6} {'実際の勝率':>12}")
    print("-" * 35)
    for idx, row in bin_summary.iterrows():
        print(f"{idx:<12} {int(row['count']):>6} {row['win_rate']:>11.2f}%")

    print("-" * 35)
else:
    print("取引データなし")

print("\n処理完了")
